### ISSUE TITLE
aws_iam_group_policy updates without name field leave orphaned attached policies

### ISSUE THREAD

**Issue by GroovyCarrot** [terraform-providers/terraform-provider-aws#4377]
### Terraform Version
Terraform v0.11.7
Provider aws 1.16.0

### Affected Resource(s)
- `aws_iam_group_policy`

Potentially also `aws_iam_role_policy` and `aws_iam_user_policy` as well, however I haven't tested yet.

### Terraform Configuration Files
```hcl
resource "aws_iam_group" "network_admins" {
  name = "NetworkAdministrators"
}

resource "aws_iam_group_policy" "network_admins" {
  group = "${aws_iam_group.network_admin.id}"
  policy = "${file("network-admins.policy.json")}"
}
```

### Expected Behavior
When a policy is updated, I expect that the policy attached to the group would be replaced.

### Actual Behavior
A new policy is attached to the group, and the old policy is left orphaned. The reason appears to be that it writes the policy with a newly generated `name`, rather than writing with the existing. Leaving `name` blank will generate i.e. `terraform-20180417120515577200000001`.

This is somewhat a security concern; if you are updating the policy to revoke permissions, since it will actually leave the old policy document attached, the group will still have the permissions you wanted to remove. You also don't have any visibility of this from Terraform.

You can get around the issue just by setting the `name` field. I have also noticed that removing `name` and running `terraform plan` shows no changes; even though I would expect it to want to replace the policy with a new computed name - this isn't really a concern though.

### Steps to Reproduce
1. Change the `policy` field - any change.
2. `terraform apply`

**Comment by bflad** [comment]
Thank you very much for reporting this. Updates were not being properly tested in the acceptance testing for generated name/name_prefix. Submitted bug fix PR with updated testing: #4379

**Comment by bflad** [comment]
This has been released in [version 1.17.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1170-may-02-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
